pkg/controller/machine.go

```diff
@@ -373,95 +333,176 @@ func (c *Controller) syncHandler(key string) error {
-				}
-				if changed {
-					glog.V(4).Infof("Updating machine '%s' with defaults...", machine.Name)
-					machine.Spec = defaultedMachineSpec
-					if machine, err = c.updateMachine(machine); err != nil {
-						return fmt.Errorf("failed to update machine '%s' after adding defaults: '%v'", machine.Name, err)
-					}
-					glog.V(4).Infof("Successfully updated machine '%s' with defaults!", machine.Name)
-				}
-				if err := c.validateMachine(prov, machine); err != nil {
-					if _, errNested := c.updateMachineError(machine, machinev1alpha1.InvalidConfigurationMachineError, err.Error()); errNested != nil {
-						return fmt.Errorf("failed to update machine error after failed validation: %v", errNested)
-					}
-					return fmt.Errorf("invalid provider config: %v", err)
-				}
-				// Remove error message in case it was set
-				if machine, err = c.clearMachineErrorIfSet(machine, machinev1alpha1.InvalidConfigurationMachineError); err != nil {
-					return fmt.Errorf("failed to update machine after removing the failed validation error: %v", err)
-				}
-				glog.V(4).Infof("Validated machine spec of %s", machine.Name)
+		err = c.createNodeForMachine(prov, providerInstance, machine, userdataProvider, providerConfig)
+		if err != nil {
+			return err
+		}
+	}
 
-				kubeconfig, err := c.createBootstrapKubeconfig(machine.Name)
-				if err != nil {
-					return fmt.Errorf("failed to create bootstrap kubeconfig: %v", err)
-				}
+	if node != nil {
+		return c.ensureNodeLabelsAnnotationsAndTaints(node, machine)
+	}
 
-				data, err := userdataProvider.UserData(machine.Spec, kubeconfig, prov, c.clusterDNSIPs)
-				if err != nil {
-					return fmt.Errorf("failed get userdata: %v", err)
-				}
+	return nil
+}
 
-				glog.Infof("creating instance...")
-				if providerInstance, err = c.createProviderInstance(prov, machine, data); err != nil {
-					if _, errNested := c.updateMachineError(machine, machinev1alpha1.CreateMachineError, err.Error()); errNested != nil {
-						return fmt.Errorf("failed to update machine error after failed machine creation: %v", errNested)
-					}
-					return fmt.Errorf("failed to create machine at cloudprovider: %v", err)
+// deleteMachineAndProviderInstance makes sure that an instance has gone in a series of steps.
+func (c *Controller) deleteMachineAndProviderInstance(prov cloud.Provider, providerInstance instance.Instance, machine *machinev1alpha1.Machine) error {
+	var err error
+
+	// step 1: delete provider instance
+	if err = c.deleteProviderInstance(prov, machine); err != nil {
+		if ok, _, _ := cloudprovidererrors.IsTerminalError(err); ok {
+			message := fmt.Sprintf("%v. Please manually delete finalizers from the machine object.", err)
+			if _, errNested := c.updateMachineError(machine, machinev1alpha1.DeleteMachineError, message); errNested != nil {
+				return fmt.Errorf("failed to update machine error after failed delete: %v", errNested)
+			}
+			return err
```

Comment:
we could also return `nil` which is quite reasonable since a `terminal` error shouldn't be requeued.